XenAdmin/Alerts/Types/LeafCoalesceAlert.cs

```diff
@@ -0,0 +1,112 @@
+﻿/* Copyright (c) Cloud Software Group, Inc. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using XenAPI;
+using static XenAPI.Message;
+
+namespace XenAdmin.Alerts
+{
+    public class LeafCoalesceAlert : MessageAlert
+    {
+        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        private MessageType LeafType;
+        private string vmID;
+
+        public LeafCoalesceAlert(Message msg)
+            : base(msg)
+        {
+            LeafType = msg.Type;
+            MapVdiToVm(msg);
+        }
+
+        private void MapVdiToVm(Message msg)
+        {            
+            var obj = msg.GetXenObject();
+            
+            if (obj is VDI vdi && vdi.Connection != null)
+            {
+                foreach (VBD vbd in vdi.Connection.ResolveAll(vdi.VBDs))

```

Comment:
I know code calling ResolveAll is used all over the application, however, a loop like this is  more effcient:
```
foreach (var vbdRef in vdi.VBDs)
{
    var vbd = vdi.Connection.Resolve(vbdRef);
    
    if (vbd != null && vbd.currently_attached)
    {
        VM vm = vbd.Connection.Resolve(vbd.VM);
        
        if (vm != null)
        {
            _vm = vm;
            break;
        }
    }
}
```
because it resolves only as many VBDss as needed (instead of all), and avoids resolving the VM for detached VDIs.